index.html

```diff
@@ -2687,13 +2687,191 @@ <h1>
 This section defines the inputs and outputs of the <a>DID resolution</a> and
 <a>DID URL dereferencing</a> processes. The exact implementation of these
 functions is out of scope for this specification, but some considerations for
-implementors are available in [[?DID-RESOLUTION]].
+implementors are discussed in [[?DID-RESOLUTION]].
+    </p>
+
+    <p>
+All conformant <a>DID resolvers</a> MUST implement the <a>DID resolution</a> function
+for at least one <a>DID method</a> and MUST be able to return a <a>DID document</a> in 
+at least one conformant representation.
     </p>
 
     <section>
         <h2>
 DID Resolution
         </h2>
+        <p>
+The <a>DID resolution</a> function resolves a <a>DID</a> into a <a>DID document</a>
+by using the "Read" operation of the applicable <a>DID method</a>. (See <a href="#read-verify"></a>.)
+The details of how this process is accomplished are outside the scope of this
+specification, but all conformant implementations MUST implement a function in the following form:
+        </p>
+
+        <p><code>
+resolve ( did, did-resolution-input-metadata ) <br>
+&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-&gt; ( did-resolution-metadata, did-document-stream, did-document-metadata )
```

Comment:
Is this syntax for these functions described somewhere?  Hard to tell looking at this if this can be an async function (in languages that do that).  A JavasScript implementation would probably return a Promise here.  Is the output a tuple or map with named properties?  Seems kind of a fuzzy description for a "MUST".  In the JSON-LD API spec we used WebIDL for these API descriptions.  Has some rough edges but seemed to work good enough and has some respec support.  The [LoadDocumentCallback](https://www.w3.org/TR/json-ld11-api/#remote-document-and-context-retrieval) API has a similar pattern to this.

Reply:
Those details depend on your language. It's pretty clear in any implementation what "call a function" and "return a value" mean, and it's different for each one. 

Here, we're not specifying if it's synchronous or not, so a promise-based resolution would be fine. We're also not specifying how multiple outputs are returned, so a tuple, or an encapsulating object, or a bunch of pointers will all be fine.

Reply:
I think it's confusing to use language like "all conformant implementations MUST implement a function in the following form".  In a spec I read this as something strict.  But here it involves guessing what the ad-hoc function call description means, and understanding it can be sync/async, use tuples, objects, or pointers (for parameters and results maybe?), and I'm guessing the actual names are not required either since many languages can't handle `-` in identifiers.  What form is it that must be followed?